safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,155 @@
+		// allowed to pass.
+		return true
+	}
+
+	if m := r.Method(); h.Get("Sec-Fetch-Mode") == "navigate" && (m == safehttp.MethodGet || m == safehttp.MethodHead) {
+		if dest := h.Get("Sec-Fetch-Dest"); dest == "object" || dest == "embed" {
+			// The request is cross-site and originates from <object> or <embed>
+			// so it is rejected.
+			return false
+		}
+		// The request is cross-site, but a simple top-level navigation so we
+		// allow it to pass.
+		return true
+	}
+
+	origin := h.Get("Origin")
+	if _, ok := allowedCORS[origin]; ok && h.Get("Sec-Fetch-Mode") == "cors" {
+		// The request is cross-site but sent from an
+		// origin where Cross-Origin Resource Sharing is explicitly allowed.
+		return true
+	}
+
+	// The request is cross-site, not navigational and sent from an endpoint
+	// where CORS is not enabled therefore it is rejected.
+	return false
+}
+
+func navigationIsolationPolicy(r *safehttp.IncomingRequest, allowedCORS map[string]bool) bool {
+	h := r.Header
+	origin := h.Get("Origin")
+	if h.Get("Sec-Fetch-Site") == "cross-site" && h.Get("Sec-Fetch-Mode") == "navigate" {
+		if _, ok := allowedCORS[origin]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// SetReportMode sets the Fetch Metadata policy mode to "report". This will
+// allow requests that violate the policy to pass, but will report the violation
+// using the LoggingService. The method will panic if no LoggingService is
+// provided.
+func (p *Plugin) SetReportMode(logger LoggingService) {
+	if logger == nil {
+		panic("logging service required for Fetch Metadata report mode")
+	}
+	p.logger = logger
+	p.reportOnly = true
+}
+
+// SetEnforceMode sets the Fetch Metadata policy mode to "enforce". This will
+// reject any requests that violate the policy provided by the plugin.
+func (p *Plugin) SetEnforceMode() {
+	p.reportOnly = false
+}
+
+// Before validates the safehttp.IncomingRequest using the Fetch Metadata policy
+// provided by the  plugin. It only allows request to pass if they conform to
+// the policy, if it's sent from a CORS endpoint, or if the mode is set to
+// "report", in which case the request is allowed to pass but the violation is
+// reported. Moreover, if the Navigation Isolation Policy is enabled, the
+// request will also be validated against it.  If the browser does not have
+// Fetch Metadata support implemented, the policy will not be applied and all
+// requests will be allowed to pass.
+func (p *Plugin) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	if !p.Policy(r, p.allowedCORS) {
+		if !p.reportOnly {
+			return w.ClientError(safehttp.StatusForbidden)
+		}
+		p.logger.Log(r)
+		return safehttp.Result{}
+	}
+	if p.NavIsolation && !navigationIsolationPolicy(r, p.allowedCORS) {
+		if !p.reportOnly {
+			return w.ClientError(safehttp.StatusForbidden)
+		}
+		p.logger.Log(r)
+		return safehttp.Result{}
+	}
+	return safehttp.Result{}
```

Comment:
A lot of these conditions are negated, and most of the important logic is indented. Moreover there is quite a bit of duplication in structure.

Can you try to refactor this to make it more linear to read?